zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+#[frame_support::pallet]
+mod pallet {
+    use crate::{
+        consts::MAX_ASSETS,
+        math::{Math, MathOps},
+        traits::{pool_operations::PoolOperations, DistributeFees, LiquiditySharesManager},
+        types::{FeeDistribution, Pool, SoloLp},
+        weights::*,
+    };
+    use alloc::collections::BTreeMap;
+    use core::marker::PhantomData;
+    use frame_support::{
+        dispatch::DispatchResultWithPostInfo,
+        ensure,
+        pallet_prelude::StorageMap,
+        require_transactional,
+        traits::{Get, IsType, StorageVersion},
+        transactional, PalletId, Twox64Concat,
+    };
+    use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+    use orml_traits::MultiCurrency;
+    use sp_runtime::{
+        traits::{AccountIdConversion, CheckedAdd, CheckedSub, Saturating, Zero},
+        DispatchError, DispatchResult, SaturatedConversion,
+    };
+    use zeitgeist_primitives::{
+        constants::{BASE, CENT},
+        math::fixed::{bdiv, bmul},
+        traits::{CompleteSetOperationsApi, DeployPoolApi, ZeitgeistAssetManager},
+        types::{Asset, MarketStatus, MarketType, ScalarPosition, ScoringRule},
+    };
+    use zrml_market_commons::MarketCommonsPalletApi;
+
+    // These should not be config parameters to avoid misconfigurations.
+    pub(crate) const MIN_SWAP_FEE: u128 = BASE / 1_000; // 0.1%.
+    pub(crate) const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+    pub(crate) const MAX_SPOT_PRICE: u128 = BASE - CENT / 2;
+    pub(crate) const MIN_SPOT_PRICE: u128 = CENT / 2;
+    pub(crate) const MIN_LIQUIDITY: u128 = BASE;
+
+    pub(crate) type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+    pub(crate) type AssetOf<T> = Asset<MarketIdOf<T>>;
+    pub(crate) type BalanceOf<T> = <<T as Config>::AssetManager as MultiCurrency<
+        <T as frame_system::Config>::AccountId,
+    >>::Balance;
+    pub(crate) type IndexType = u32;
+    pub(crate) type MarketIdOf<T> =
+        <<T as Config>::MarketCommons as MarketCommonsPalletApi>::MarketId;
+    pub(crate) type PoolOf<T> = Pool<T, SoloLp<T>>;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type AssetManager: ZeitgeistAssetManager<Self::AccountId, CurrencyId = AssetOf<Self>>;
+
+        type CompleteSetOperations: CompleteSetOperationsApi<
+                AccountId = Self::AccountId,
+                Balance = BalanceOf<Self>,
+                MarketId = MarketIdOf<Self>,
+            >;
+
+        /// Distribute external fees. The fees are paid from the pool account, which in turn has
+        /// received the fees from the trader.
+        type ExternalFees: DistributeFees<
+                Asset = AssetOf<Self>,
+                AccountId = AccountIdOf<Self>,
+                Balance = BalanceOf<Self>,
+                MarketId = MarketIdOf<Self>,
+            >;
+
+        type MarketCommons: MarketCommonsPalletApi<AccountId = Self::AccountId, BlockNumber = Self::BlockNumber>;
```

Comment:
Yes, it absolutely makes sense. Some bit of coupling will have to remain because we need the market ID to perform complete set operations. Will have to be done anyways once we go full combo. However, note that we are currently enjoying some advantages by coupling the two concepts: Right now, we don't have to deal with stuff like _pool status_ - we just piggyback on the market's status. We don't need automated opening/closing of pools, etc.